src/main/java/org/logstash/filters/GeoIpFilter.java

```diff
@@ -0,0 +1,343 @@
+    Event event = rubyEvent.getEvent();
+    Object input = event.getField(sourceField);
+    if (input == null) {
+      return false;
+    }
+
+    String ip;
+
+    if (input instanceof List) {
+      ip = (String) ((List) input).get(0);
+    } else if (input instanceof String) {
+      ip = (String) input;
+    } else {
+      throw new IllegalArgumentException("Expected input field value to be String or List type");
+    }
+
+    Map<String, Object> geoData = new HashMap<>();
+
+    try {
+      final InetAddress ipAddress = InetAddress.getByName(ip);
+      switch (databaseReader.getMetadata().getDatabaseType()) {
+        case CITY_DB_TYPE:
+          geoData = retrieveCityGeoData(ipAddress);
+          break;
+        case COUNTRY_DB_TYPE:
+          geoData = retrieveCountryGeoData(ipAddress);
+          break;
+        case ISP_DB_TYPE:
+          geoData = retrieveIspGeoData(ipAddress);
+          break;
+        default:
+          throw new IllegalStateException("Unsupported database type " + databaseReader.getMetadata().getDatabaseType() + "");
+      }
+    } catch (UnknownHostException e) {
+      logger.debug("IP Field contained invalid IP address or hostname. exception={}, field={}, event={}", e, sourceField, event);
+    } catch (AddressNotFoundException e) {
+      logger.debug("IP not found! exception={}, field={}, event={}", e, sourceField, event);
+    } catch (GeoIp2Exception | IOException e) {
+      logger.debug("GeoIP2 Exception. exception={}, field={}, event={}", e, sourceField, event);
+    }
+
+    return applyGeoData(geoData, event);
+  }
+
+  private boolean applyGeoData(Map<String, Object> geoData, Event event) {
+    if (geoData == null) {
+      return false;
+    }
+    // only do event.set(@target) if the lookup result is not null
+    if (event.getField(targetField) == null) {
+      event.setField(targetField, Collections.emptyMap());
+    }
+    // don't do anything more if the lookup result is empty
+    if (geoData.isEmpty()) {
+      return false;
+    }
+
+    String s = "[" + this.targetField + "][";
+    for (Map.Entry<String, Object> it: geoData.entrySet()) {
+      event.setField(s + it.getKey() + "]", it.getValue());
+    }
+    return true;
+  }
+
+  private Map<String,Object> retrieveCityGeoData(InetAddress ipAddress) throws GeoIp2Exception, IOException {
+    CityResponse response = databaseReader.city(ipAddress);
+    Country country = response.getCountry();
+    City city = response.getCity();
+    Location location = response.getLocation();
+    Continent continent = response.getContinent();
+    Postal postal = response.getPostal();
+    Subdivision subdivision = response.getMostSpecificSubdivision();
+    Map<String, Object> geoData = new HashMap<>();
+
+    // if location is empty, there is no point populating geo data
+    // and most likely all other fields are empty as well
+    if (location.getLatitude() == null && location.getLongitude() == null) {
```

Comment:
Since you're rewriting this anyway, please don't keep this mistake. [There are GeoIP records with no lat-lon](https://github.com/logstash-plugins/logstash-filter-geoip/issues/107#issuecomment-292182214) that this plugin should be able to handle.